mlflow/autogen/autogen_logger.py

```diff
@@ -0,0 +1,262 @@
+import logging
+import time
+import uuid
+from dataclasses import dataclass, field
+from datetime import datetime
+from typing import Any, Callable, Dict, List, Optional, Union
+
+from autogen import Agent, ConversableAgent
+from autogen.logger.base_logger import BaseLogger
+from openai.types.chat import ChatCompletion
+
+from mlflow import MlflowClient
+from mlflow.entities.span import NoOpSpan, Span, SpanType
+from mlflow.entities.span_event import SpanEvent
+from mlflow.entities.span_status import SpanStatus, SpanStatusCode
+from mlflow.tracing.utils import capture_function_input_args
+
+# For GroupChat, a single "received_message" events are passed around multiple
+# internal layers and thus too verbose if we show them all. Therefore we ignore
+# some of the message senders listed below.
+_EXCLUDED_MESSAGE_SENDERS = ["chat_manager", "checking_agent"]
+
+_logger = logging.getLogger(__name__)
+
+
+@dataclass
+class ChatState:
+    """
+    Represents the state of a chat session.
+    """
+
+    # The root span object that scopes the entire single chat session. All spans
+    # such as LLM, function calls, in the chat session should be children of this span.
+    session_span: Optional[Span] = None
+    # The last message object in the chat session.
+    last_message: Optional[Any] = None
+    # The timestamp (ns) of the last message in the chat session.
+    last_message_timestamp: int = 0
+    # LLM/Tool Spans created after the last message in the chat session.
+    # We consider them as operations for generating the next message and
+    # re-locate them under the corresponding message span.
+    pended_spans: List[Span] = field(default_factory=list)
+
+    def clear(self):
+        self.session_span = None
+        self.last_message = None
+        self.last_message_timestamp = 0
+        self.pended_spans = []
+
+
+class MlflowAutogenLogger(BaseLogger):
+    def __init__(self):
+        self._client = MlflowClient()
+        self._chat_state = ChatState()
+
+    def start(self) -> str:
+        return "session_id"
+
+    def log_new_agent(self, agent: ConversableAgent, init_args: Dict[str, Any]) -> None:
+        """
+        This handler is called whenever a new agent instance is created.
+        Here we patch the agent's methods to start and end a trace around its chat session.
+        """
+        if hasattr(agent, "initiate_chat"):
+            original = agent.initiate_chat
+            # Setting root_only = True because sometimes compounded agent calls initiate_chat()
+            # method of its sub-agents, which should not start a new trace.
+            agent.initiate_chat = self._patch_function_call(
+                original, "initiate_chat", root_only=True
+            )
+        # TODO: Patch generate_reply() method as well
```

Comment:
Do we want to patch the `a_initiate_chat` entry point as well?

Reply:
Great point. Let me address it in a separate PR. It seems the current design of pending-then-resolve work with async method because the chat messages are created sequentially anyway ([code](https://github.com/microsoft/autogen/blob/2a436c90e6e55bd3af07abbcd8ea7da1870134ec/autogen/agentchat/conversable_agent.py#L1136-L1146)).